bolt-modules/boltlib/lib/puppet/functions/parallelize.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+require 'bolt/yarn'
+
+# Accepts an array and a block and maps the block to the array in parallel
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:parallelize, Puppet::Functions::InternalFunction) do
+  # Map a block to an array in parallel
+  # @param data The array to apply the block to
+  # @return [Array] An array of Result or Error objects from the block for each array element
+  # @example Execute 2 tasks on multiple targets. Once the task finishes on one
+  #   target, that target can move to the next step without waiting for the task
+  #   to finish on the second target
+  # $targets = get_targets(["host1", "host2"])
+  # $result = parallelize ($targets) |$t| {
+  #   run_task('b', $t)
+  #   run_task('c', $t)
+  # }
+  dispatch :parallelize do
+    scope_param
+    param 'Array[Any]', :data
+    block_param 'Callable[Any]', :block
+    return_type 'Array[Boltlib::PlanResult]'
+  end
+
+  def parallelize(scope, data, &block)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'parallelize')
+    end
+
+    executor = Puppet.lookup(:bolt_executor)
+    executor.report_function_call(self.class.name)
+
+    skein = data.each_with_index.map do |object, index|
+      fiber = Fiber.new do
```

Comment:
I think it would make sense to have an executor method that does basically everything inside this block (the equivalent of the former `future()` function).